### PR TITLE
xborders: init at 3.4

### PR DESCRIPTION
**Pull request by ElnuDev**
###### Description of changes

Init xborders at 3.4. xborders (https://github.com/deter0/xborder) is a program that provides an active window border replacement for window managers. Despite the repository being called xborder not xborders and inconsistent naming, the owner has confirmed that the name is https://github.com/deter0/xborder/issues/36#issuecomment-1244355598. In `preBuild`, I'm replacing the contents of the `get_version` function with the version number, because otherwise xborders requires a version.txt file to be next to the executable to fetch the version number from.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/window-managers/xborders/default.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
+{ lib, python3Packages, fetchFromGitHub, libwnck, gtk3, libnotify, wrapGAppsHook, gobject-introspection }:
+
+with python3Packages; buildPythonPackage rec {
```
Comment (AndersonTorres) [inline pkgs/applications/window-managers/xborders/default.nix:3]:
```suggestion
buildPythonPackage rec {
```

Do not abuse `with`.

File: pkgs/applications/window-managers/xborders/default.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
+{ lib, python3Packages, fetchFromGitHub, libwnck, gtk3, libnotify, wrapGAppsHook, gobject-introspection }:
```
Comment (AndersonTorres) [inline pkgs/applications/window-managers/xborders/default.nix:1]: One per line.

File: pkgs/applications/window-managers/xborders/default.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+with python3Packages; buildPythonPackage rec {
+  pname = "xborders";
+  version = "3.4"; # in version.txt
```
Comment (AndersonTorres) [inline pkgs/applications/window-managers/xborders/default.nix:5]:
```suggestion
  version = "3.4"; # in version.txt
  
```

File: pkgs/applications/window-managers/xborders/default.nix [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  src = fetchFromGitHub {
+    owner = "deter0";
+    repo = "xborder";
+    rev = "d840ed649e92b7ad399cdb1a30aeca7d09f2c08f";
+    sha256 = "sha256-Q2WWXekUL/OOr/OBqJOZbAHaWURbP+coBVp206uPSNY=";
+  };
```
Comment (AndersonTorres) [inline pkgs/applications/window-managers/xborders/default.nix:11]:
```suggestion
  };

```

File: pkgs/applications/window-managers/xborders/default.nix [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  buildInputs = [ libwnck gtk3 libnotify ];
+  nativeBuildInputs = [ wrapGAppsHook gobject-introspection ];
+  propagatedBuildInputs = [ pycairo requests pygobject3 ];
```
Comment (AndersonTorres) [inline pkgs/applications/window-managers/xborders/default.nix:14]:
```suggestion
  buildInputs = [ libwnck gtk3 libnotify ];
  
  nativeBuildInputs = [ wrapGAppsHook gobject-introspection ];
  
  propagatedBuildInputs = [ pycairo requests pygobject3 ];
  
```

File: pkgs/applications/window-managers/xborders/default.nix [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  preBuild = ''
+    sed '52,58!d' ${pname}
+    sed -i '51 a\    return ${version}' ${pname}
```
Comment (AndersonTorres) [inline pkgs/applications/window-managers/xborders/default.nix:17]: Do not abuse `${pname}`. Indeed, what these lines do?

File: pkgs/applications/window-managers/xborders/default.nix [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    cat >setup.py <<'EOF'
```
Comment (AndersonTorres) [inline pkgs/applications/window-managers/xborders/default.nix:18]: Shouldn't it be upstreamed?

File: pkgs/top-level/all-packages.nix [hunk 8]
```diff
@@ -33956,6 +33956,8 @@ with pkgs;
     gtk = gtk2;
   };
 
+  xborders = callPackage ../applications/window-managers/xborders { };
```
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:33959]: This is not a window manager. Throw it under `tools/X11`.

File: pkgs/applications/window-managers/xborders/default.nix [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    )
+    EOF
+  '';
+  meta = with lib; {
+    description = "Active window border replacement for window managers.";
+    homepage = "https://github.com/deter0/xborder";
+    license = licenses.unlicense;
+    maintainers = with maintainers; [ elnudev ];
+    platforms = platforms.linux;
+  };
```
Comment (AndersonTorres) [inline pkgs/applications/window-managers/xborders/default.nix:38]:
```suggestion
  meta = {
    description = "Active window border replacement for window managers.";
    homepage = "https://github.com/deter0/xborder";
    license = lib.licenses.unlicense;
    maintainers = with lib.maintainers; [ elnudev ];
    platforms = lib.platforms.linux;
  };
```

File: pkgs/applications/window-managers/xborders/default.nix (hunk 5 above)
Comment (AndersonTorres) [reply pkgs/applications/window-managers/xborders/default.nix:14]: Also, one per line.

File: pkgs/applications/window-managers/xborders/default.nix [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  src = fetchFromGitHub {
+    owner = "deter0";
+    repo = "xborder";
+    rev = "d840ed649e92b7ad399cdb1a30aeca7d09f2c08f";
+    sha256 = "sha256-Q2WWXekUL/OOr/OBqJOZbAHaWURbP+coBVp206uPSNY=";
+  };
+  buildInputs = [ libwnck gtk3 libnotify ];
+  nativeBuildInputs = [ wrapGAppsHook gobject-introspection ];
+  propagatedBuildInputs = [ pycairo requests pygobject3 ];
+  preBuild = ''
```
Comment (AndersonTorres) [inline pkgs/applications/window-managers/xborders/default.nix:15]: I should be a `postPhase`.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/applications/window-managers/xborders/default.nix (hunk 6 above)
Comment (ElnuDev) [reply pkgs/applications/window-managers/xborders/default.nix:17]: I mentioned what these do in change description.

> In `preBuild`, I'm replacing the contents of the `get_version` function with the version number, because otherwise xborders requires a version.txt file to be next to the executable to fetch the version number from.

Basically, the author does this really weird thing where the version number is stored in [version.txt](https://github.com/deter0/xborder/blob/d840ed649e92b7ad399cdb1a30aeca7d09f2c08f/version.txt#L1) and then is fetched in the [`get_version` method](https://github.com/deter0/xborder/blob/d840ed649e92b7ad399cdb1a30aeca7d09f2c08f/xborders#L51-L58). version.txt has to be next to the executable Python file, which is problematic since I don't think it'd be a good idea to have a data file in the `$out/bin` folder. It might also cause issues with the wrapping, I'm not sure.

All these lines do is replace the contents of `get_version` with a single return, removing the dependency of version.txt:

```PY
def get_version():
    return 3.4
```

File: pkgs/top-level/all-packages.nix (hunk 8 above)
Comment (ElnuDev) [reply pkgs/top-level/all-packages.nix:33959]: Alright. I noticed some things that were window manager-tangential such as picom were in `applications/window-managers`, so I thought it'd be appropriate to put it there since it's related. I'll move it.

File: pkgs/applications/window-managers/xborders/default.nix (hunk 9 above)
Comment (ElnuDev) [reply pkgs/applications/window-managers/xborders/default.nix:38]: Is this necessary? In my previous packaging PRs this wasn't an issue, it seems overly verbose without the `with lib;`. I think most of the package derivations in nixpkgs I've looked at so far do it this way.

File: pkgs/applications/window-managers/xborders/default.nix (hunk 7 above)
Comment (ElnuDev) [reply pkgs/applications/window-managers/xborders/default.nix:18]: Maybe? I'll make a PR upstream and see what the owner says.

File: pkgs/applications/window-managers/xborders/default.nix (hunk 10 above)
Comment (ElnuDev) [reply pkgs/applications/window-managers/xborders/default.nix:15]: You mean `preBuild` should be `postPhase`? It doesn't build with `postPhase` because `setup.py` needs to be created before building, otherwise `buildPythonPackage` fails.

File: pkgs/applications/window-managers/xborders/default.nix (hunk 6 above)
Comment (AndersonTorres) [reply pkgs/applications/window-managers/xborders/default.nix]: Well, the first obvious problem is it is strongly depends on the position of the code.
If by any reason the upstream modifies the lines, your sed script will destroy the file without warning.

Change this to an explicit `.diff` file instead.

File: pkgs/applications/window-managers/xborders/default.nix (hunk 7 above)
Comment (ElnuDev) [reply pkgs/applications/window-managers/xborders/default.nix]: https://github.com/deter0/xborder/pull/55

File: pkgs/applications/window-managers/xborders/default.nix (hunk 9 above)
Comment (AndersonTorres) [reply pkgs/applications/window-managers/xborders/default.nix]: > Is this necessary? 

No. The code will not melt the computer if you don't do this.

On the other hand, there are some arguments against `with` itself as a bad language construction. The main arguments are compiled [here](https://nix.dev/anti-patterns/language#with-attrset-expression).

The text above is too radical for some, but the nested `with` is worth noticing:

https://github.com/NixOS/nix/issues/490

https://github.com/pikajude/nixpkgs/commit/eea78e554e03906524f87406bb3a17d5c5264029

> In my previous packaging PRs this wasn't an issue

Yes. It happens to everyone.

Me included - some old expressions I wrote contain much of it, because, let's be honest, almost everyone copy-pastes those blocks.

> it seems overly verbose without the `with lib;`.

"Excessive verbosity" was never a sufficient reason to start every derivation with

```nix
{ pkgs }: with pkgs;
mkDerivation { . . . }
```

> I think most of the package derivations in nixpkgs I've looked at so far do it this way.

Yes, and it should be changed.

File: pkgs/applications/window-managers/xborders/default.nix (hunk 10 above)
Comment (AndersonTorres) [reply pkgs/applications/window-managers/xborders/default.nix]: Sorry, I meant `postPatch`. `preBuild` should be `postPatch` because, semantically, it is a patch - you are modifying the source files.

Also, you should include this `setup.py` file as a patch too - or at least directly as a file, not as a heredoc.

File: pkgs/applications/window-managers/xborders/default.nix (hunk 9 above)
Comment (ElnuDev) [reply pkgs/applications/window-managers/xborders/default.nix]: I understand the issue with having a `with` at the beginning of the derivation, but I'm still a bit confused what the problem is with using `with lib;` in `meta`. It's only a small set and almost all of the properties' values are inside `lib`, so it doesn't seem to me like there would be any downsides of using it there. What's the advantage of omitting it?

I'll make the changes.

File: pkgs/tools/X11/xborders/default.nix [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  lib,
+  buildPythonPackage,
+  python3Packages,
+  fetchFromGitHub,
+  libwnck,
+  gtk3,
+  libnotify,
+  wrapGAppsHook,
+  gobject-introspection
+}:
```
Comment (AndersonTorres) [inline pkgs/tools/X11/xborders/default.nix:11]: Are you using [Alejandra](https://github.com/kamadorueda/alejandra)? 
According to https://github.com/NixOS/rfcs/pull/101, the blessed code formatter will be [nixpkgs-fmt](https://github.com/nix-community/nixpkgs-fmt). 

```suggestion
{ lib
  , buildPythonPackage
  , python3Packages
  , fetchFromGitHub
  , libwnck
  , gtk3
  , libnotify
  , wrapGappsHook
  , gobject-introspection
}:
```

Comment (ElnuDev) [reply pkgs/tools/X11/xborders/default.nix:11]: Oh I didn't know that was a thing! I'll use it in future, thanks.

File: pkgs/applications/window-managers/xborders/default.nix (hunk 9 above)
Comment (AndersonTorres) [reply pkgs/applications/window-managers/xborders/default.nix]: If the example in libuv quoted above - and below, again

https://github.com/pikajude/nixpkgs/commit/eea78e554e03906524f87406bb3a17d5c5264029

is not convincing, and you believe there is absolutely nothing wrong in code like

```nix
   platforms = with platforms; linux ++ darwin;
```

then I have no more educational devices to rely on. Copy-paste as you wish.

File: pkgs/applications/window-managers/xborders/default.nix (hunk 7 above)
Comment (ElnuDev) [reply pkgs/applications/window-managers/xborders/default.nix]: Merged.

**Comment by ElnuDev** [comment]
I'm not really sure how to go about having setup.py as a patch. The other patch code has been removed since the version.txt has been fixed upstream.

**Comment by ElnuDev** [comment]
I'm a little bit confused whether or not I'm doing something wrong here. When you run `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`, the build outputs should be in `./result/bin`, right? For some reason there's nothing there, and it doesn't say anything was built in the report files. But ofborg doesn't seem to be giving any errors? This derivation is working in my personal config.

File: pkgs/tools/X11/xborders/default.nix [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+  ];
+
+  propagatedBuildInputs = with python3Packages; [
+    pycairo
+    requests
+    pygobject3
+  ];
+
+  postPatch = ''
+    cat >setup.py <<'EOF'
```
Comment (AndersonTorres) [inline pkgs/tools/X11/xborders/default.nix:41]: Why is this file needed to begin with?

I would ask if how the other distros are packaging this, but I have found none.

Supposing the file is needed:

1. Open a shell;
2. `cd local-nixpkgs-directory/pkgs/tools/X11/xborders`;
3. `nano config.py` -- insert contents to it as needed
4. `nano default.nix` -- insert something like

```nix
postPatch = ''
  ln -s ${./setup.py} setup.py
'';
```

Comment (ElnuDev) [reply pkgs/tools/X11/xborders/default.nix:41]: Ah alright, I was thinking one wanted to include `setup.py` as an extra source rather than just being linked in with a script. In terms of getting variable substitution work for the description, etc. I guess one can use `substituteAll`?

`setup.py` is required for `buildPythonPackage` (it fails without it), and the documentation for fixing a missing `setup.py` included shows including it in `preBuild`, which is why I had it that way originally: [Fix Missing `setup.py`](https://nixos.wiki/wiki/Packaging/Python#Fix_Missing_setup.py)

Comment (AndersonTorres) [reply pkgs/tools/X11/xborders/default.nix:41]: Understood, but not much.

The arguments and example from the wiki are cogent but they lack elegance.
It is better to create a separate file `setup.py` and move it instead of embedding it in a heredoc in a shell script.

About using `substitute`, I believe this is ugly and unneeded.

Comment (ElnuDev) [reply pkgs/tools/X11/xborders/default.nix]: Alright. I agree that using `substituteAll` is ugly but I feel like having two update two files for a version bump would be worse. I've added the changes.

File: pkgs/tools/X11/xborders/default.nix [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+      src = ./setup.py;
+      desc = meta.description; # "description" is reserved
+      inherit pname version;
+    };
+  in ''
+    ln -s ${setup} setup.py
+  '';
+
+  meta = with lib; {
+    description = "Active window border replacement for window managers.";
```
Comment (AndersonTorres) [inline pkgs/tools/X11/xborders/default.nix:52]:
```suggestion
    description = "Active window border replacement for window managers";
```

File: pkgs/tools/X11/xborders/default.nix [hunk 14]
```diff
@@ -0,0 +1,58 @@
+{ lib
+  , buildPythonPackage
+  , python3Packages
+  , fetchFromGitHub
+  , libwnck
+  , gtk3
+  , libnotify
+  , wrapGAppsHook
+  , gobject-introspection
+  , substituteAll
```
Comment (AndersonTorres) [inline pkgs/tools/X11/xborders/default.nix:10]:
```suggestion
, fetchFromGitHub
, python3Packages
, gobject-introspection
, gtk3
, libnotify
, libwnck
, substituteAll
, wrapGAppsHook
```

File: pkgs/tools/X11/xborders/default.nix [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+}:
+
+buildPythonPackage rec {
```
Comment (AndersonTorres) [inline pkgs/tools/X11/xborders/default.nix:13]:
```suggestion
python3Packages.buildPythonPackage rec {
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Otherwise lgtm

**Comment by ElnuDev** [comment]
Everything seems okay! I've checked the build results from `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"` and it works.

Thanks for all the help and reviews with this PR, I really appreciate it.
